### PR TITLE
Add `Data.Vinyl.Curry`

### PR DESCRIPTION
**Pull request by bch29**
I find these combinators very useful when using Vinyl records as heterogeneous lists.

### CODE REVIEW

**Comment by acowley** [comment]
I use this technique in another project, too, so I'm enthusiastic about adding it here. Do you think a small example of usage would fit in the module comment?

I'd like to have something somewhere showing usage because the type family in the type signatures is a bit opaque. I typically use the `runcurry` variants in my own stuff; if that's also the case for you, the example should focus on that.

I'm also happy to write some documentation or examples if you'd prefer. 

Oh, another option that could fit very well here is a doctest or two. What do you think? I want to make sure people get the idea of what's offered here as it's so useful.

**Comment by bch29** [comment]
Here's some more documentation with some examples. Combinators like `rcurry` and `runcurry` (i.e. the ones that work with `Rec f ts` for general `f`) don't really do much useful with any of the `base` functors (or really anything of kind `* -> *`), so it's hard to think of simple examples for them.

I've fixed the build on GHC 7.10.3 too (I think).

Doctests do sound useful, but I'm not sure how to get them working.

**Comment by bch29** [comment]
Thoughts on the name of `rliftComposeA`? It feels like too much of a mouthful to me.

Also, do you think it might be better to make `Curried` and `CurriedF` closed type families, separate from the type class? It does currently feel a little strange that `runcurry` etc rely on those type families, without an accompanying `RecordCurry` constraint.

**Comment by bch29** [comment]
> Also, do you think it might be better to make `Curried` and `CurriedF` closed type families, separate from the type class?

I've decided to make the change, partly because I'm not sure if it was otherwise possible to make `CurriedF` poly-kinded without adding a kind parameter to the `RecordCurry` class.

File: Data/Vinyl/Curry.hs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+instance RecordCurry ts => RecordCurry (t ': ts) where
+  rcurry f x = rcurry (\xs -> f (x :& xs))
+  {-# INLINABLE rcurry #-}
+  rcurry' f x = rcurry' (\xs -> f (Identity x :& xs))
+  {-# INLINABLE rcurry' #-}
+
+-- * Uncurrying
+
+-- | N-ary version of 'uncurry' over functorial records.
```
Comment (Lysxia) [inline Data/Vinyl/Curry.hs]: Some examples of the specialized type of `rcurry` and `runcurry` on small records would help those unfamiliar with type families.

> runcurry :: (f x -> f y -> a) -> Rec f '[x, y] -> a
> -- etc.

**Comment by acowley** [comment]
Thank you, this is looking even nicer now. The name `rliftComposeA` is a mouthful, but I find most uses of `Compose` to be so. 

@Lysxia suggests even more specialization examples could help, which is hard to argue against. The examples you have are the right size, in my opinion, so perhaps carrying them through more pervasively is a good move.

Let’s sort that out and leave a few more hours at least in case anyone has a bright idea about the naming.

**Comment by bch29** [comment]
Good points all round. Any more suggestions?

**Comment by bch29** [comment]
I'm thinking `rliftA` and `rliftComposeA` might be better moved to `Data.Vinyl.Core` since they don't directly have anything to do with currying.

**Comment by acowley** [comment]
Agreed. I feel like I’ve defined this somewhere else, but I’m looking around and can’t find where.

Something that is confusing my search is that I usually name this something based on the word `fold`. I’ll keep digging around. `Curry` is not the worst home for it since writing the fold function will often use `runcurry`.

**Comment by bch29** [comment]
Ah, I've just realised that `rliftComposeA f = fmap f . rtraverse getCompose`. Perhaps we'd be better off defining

```haskell
rsequence :: (Applicative f) => Rec (Compose f g) ts -> f (Rec g ts)
rsequence = rtraverse getCompose
```

then people can use `fmap f . rsequence` and we can do away with `rliftA` and `rliftComposeA`.

**Comment by bch29** [comment]
So, anything you think needs changing before this is mergeable?

**Comment by acowley** [comment]
Nope. Thanks for the ping, I wanted to see if any other watchers might chime in. Thanks for doing such a great job working this PR into shape!

**Pull request merged**